docs/language-reference/04-types.md

```diff
@@ -0,0 +1,323 @@
+Types
+=====
+
+This section defines the kinds of types supported by Slang.
+
+Types in Slang do not necessarily prescribe a single _layout_ in memory.
+The discussion of each type will specify any guarantees about layout it provides; any details of layout not specified here may depend on the target platform, compiler options, and context in which a type is used.
+
+Void Type
+---------
+
+The type `void` contains no data and has a single, unnamed, value.
+
+A `void` value takes up no space, and thus does not affect the layout of types.
+Formally, a `void` value behaves as if it has a size of zero bytes, and one-byte alignment.
+
+Scalar Types
+------------
+
+### Boolean Type
+
+The type `bool` is used to represent Boolean truth value: `true` and `false`.
+
+The size of a `bool` varies across target platforms; programs that need to ensure a matching in-memory layout between targets should not use `bool` for in-memory data structures.
+On all platforms, the `bool` type must be _naturally aligned_ (its alignment is its size).
+
+### Integer Types
+
+The following integer types are defines:
+
+| Name          | Description |
+|---------------|-------------|
+| `int8_t`      | 8-bit signed integer |
```

Comment:
Probably, and adding them as `typedef`s should be too hard.

One thing we need to decide is what the "real" names of all these types should be, and then what should be the type aliases.

Reply:
One issue with adding `int32_t` as a `typedef` of `int` (or vice versa) is that it closes the door to us ever having `int` be a pointer/word-sized integer on each target platform. I suppose that was never going to fly given how ingrained (`int` == 32 bits) is for most of our target users.

Eventually we might want to have a distinct `intptr_t` type in the table here, sitting between `int32_t` and `int64_t`, so that conversions of 32- and 64-bit integers to/from pointer-sized integers are always explicit in the AST and initial IR (since our initial IR is supposed to be portable across targets).